Migrate NSURLCredential from WKKeyedCoder

sheeparegreat opened the pull request:
#### 49bf33d6d4354a95f5abccfe7c2361f5aebc124d
<pre>
Migrate NSURLCredential from WKKeyedCoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=282456">https://bugs.webkit.org/show_bug.cgi?id=282456</a>
<a href="https://rdar.apple.com/135824840">rdar://135824840</a>

Reviewed by Sihui Liu.

This change extracts NSURLCredential information for
Password, ServerTrust, Kerberos, MobileMe and OAuth2.
Client certificates are handled via AuthenticationManagerCocoa.mm

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDate.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
(WebKit::CoreIPCNSURLCredential::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
* Source/WebKit/Shared/cf/CoreIPCSecTrust.h:
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, Basic)):

Canonical link: <a href="https://commits.webkit.org/286450@main">https://commits.webkit.org/286450@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5de4ec28d847241cfa3de0a24ef4320062124c99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28723 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80318 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27092 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77942 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3149 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59453 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17618 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65121 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39812 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46726 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22604 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25419 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67843 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22942 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81787 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2005 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67825 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65088 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66989 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16729 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10932 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9060 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3145 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5951 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3166 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h:
```diff
@@ -25,14 +25,63 @@
+};
+
+struct CoreIPCNSURLCredentialData {
+
```
unnecessary blank line

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h:
```diff
@@ -25,14 +25,63 @@
+    using Flags = std::pair<CoreIPCString, CoreIPCString>;
+    using Attributes = std::pair<CoreIPCString, std::variant<CoreIPCNumber, CoreIPCString, CoreIPCDate>>;
+
+    CoreIPCNSURLCredentialPersistence persistence;
```
I think we need default value for this.

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,214 @@
+}
+
+CoreIPCNSURLCredential::CoreIPCNSURLCredential(CoreIPCNSURLCredentialData&& data)
+: m_data(WTFMove(data)) { }
```
Indentation?

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,214 @@
+        [dict setObject:(*m_data.user).toID().get() forKey:@"user"];
+        [dict setObject:(*m_data.password).toID().get() forKey:@"password"];
+        if (m_data.attributes) {
+            auto attributes = adoptNS([[NSMutableDictionary alloc] initWithCapacity:(*m_data.attributes).size()]);
```
`RetainPtr`; we prefer explicit type nowadays

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,214 @@
+    [dict setObject:persistence.get() forKey:@"persistence"];
+
+    if (m_data.user && m_data.password) {
+        [dict setObject:@(kURLCredentialInternetPassword) forKey:@"type"];
```
It's a bit weird that we decide `type` value based on what members are available. Can't we just store the type in `CoreIPCNSURLCredentialData`? Then here we could check corresponding members.
E.g. 
```
if (m_data.type == kURLCredentialInternetPassword) {
    // Assert  m_data.user and m_data.password is available
    // Construct dict as needed 
}
```

sheeparegreat replied:
The members which are set let us know which type it is. We could switch on a value in the struct but that is coming from an attacker controlled WebContent Process.

szewai replied:
hmm the other parameters in `CoreIPCNSURLCredentialData` are also extracted from IPC message, right? Why can we trust the other parameters but not type?
Also, can't we could cross check type and presence of other parameters to know whether the value is right?

sheeparegreat replied:
For deserialization we cannot trust any of the values. We could switch on an enum and then validate or we can infer it by the items in the struct we have received.

sheeparegreat replied:
(For NSURLCredential the flow would be NP -> UIP, not WCP)

szewai replied:
> or we can infer it by the items in the struct we have received.

I don't think we should infer it when it can be explicitly specified by sender. I noticed the weirdness because the existing implementation actually resets "type" and "realm" properties at `if (m_data.appleID && m_data.realm && m_data.password)` and `if (m_data.token && m_data.realm)`. 

If the sender process sets all these properties (no matter it is under attack or there is a bug), we won't get the type right. We could fix the issue with `else if (m_data.token && m_data.realm)`, but it still seems fragile since the process could still set all properties and the type result will be based on which type the code checks first.

So I think a more robust way is to handle properties based on the specified type, and we should only handle it once (if the properties do not match specified type, we just fail instead of trying to match next type).

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,239 @@
+    if (secTrust && CFGetTypeID(secTrust) == SecTrustGetTypeID())
+        m_data.trust = CoreIPCSecTrust(secTrust);
+    else
+        m_data.trust = CoreIPCSecTrust(nil);
```
Is this necessary? I think `m_data.trust` should be null if not set

sheeparegreat replied:
in CoreIPCNSURLCredentialData  trust is currently a CoreIPCSecTrust trust.  On the deserialization side we currently have:

```
    case CoreIPCNSURLCredentialType::ServerTrust: {
        RetainPtr<SecTrustRef> trust = m_data.trust.createSecTrust();
        if (trust) {
            [dict setObject:@(kURLCredentialServerTrust) forKey:@"type"];
            [dict setObject:(id)trust.get() forKey:@"trust"];
        }
        break;
    }
```

sheeparegreat replied:
This can probably be left default constructed.

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,239 @@
+    SET_MDATA(token, NSString, CoreIPCString);
+
+    if (m_data.user && m_data.password)
+        m_data.type = CoreIPCNSURLCredentialType::Password;
```
Can we get type from dict with key "type" directly? This seems should be returned by the SPI since we are writing that property for the dict in  `- (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;`

szewai commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
```diff
@@ -26,9 +26,249 @@
+
+    switch (m_data.type) {
+    case CoreIPCNSURLCredentialType::Password:
+        if (m_data.user && m_data.password) {
```
We should not need to check the existence of members now that type is not inferred. If we want to be sure user and password exist, we could add assertion for them. Like:
```
[dict setObject:@(kURLCredentialInternetPassword) forKey:@"type"];
ASSERT(m_data.user);
if (m_data.user)
    [dict setObject:(*m_data.user).toID().get() forKey:@"user"];
ASSERT(m_data.password);
if (m_data.password)
    [dict setObject:(*m_data.password).toID().get() forKey:@"password"];
...
if (m_data.useKeychain)
    [dict setObject:[NSNumber numberWithBool:(*m_data.useKeychain)] forKey:@"useKeychain"];
break;
```

sheeparegreat committed:
Migrate NSURLCredential from WKKeyedCoder
https://bugs.webkit.org/show_bug.cgi?id=282456
rdar://135824840

Reviewed by Sihui Liu.

This change extracts NSURLCredential information for
Password, ServerTrust, Kerberos, MobileMe and OAuth2.
Client certificates are handled via AuthenticationManagerCocoa.mm

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDate.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm:
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
(WebKit::CoreIPCNSURLCredential::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
* Source/WebKit/Shared/cf/CoreIPCSecTrust.h:
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, Basic)):

Canonical link: https://commits.webkit.org/286450@main
